File: Arrays/intersection-of-two-arrays/intersection_of_two_arrays.py
#  Project : SDE Placement Questions
#  Filename : intersection_of_two_arrays.py
#  Author : thameem
#  Current modification time : Sun, 23 Oct 2022 at 11:06 pm India Standard Time
#  Last modified time : Sun, 23 Oct 2022 at 11:06 pm India Standard Time


def intersection(nums1: list[int], nums2: list[int]) -> list[int]:
    nums1 = list(set(nums1))
    nums2 = list(set(nums2))

    _intersection = set()

    for num in nums1:
        if num in nums2:
            _intersection.add(num)

    return list(_intersection)


# todo: optimize with n+m


if __name__ == '__main__':
    print(intersection(nums1=[1, 2, 2, 1], nums2=[2, 2]))

File: Arrays/intersection-of-two-arrays/intersection_of_two_arrays.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <set>

using namespace std;

vector<int> intersection(vector<int> &nums1, vector<int> &nums2)
{
    vector<int> _intersection;
    set<int> n1,n2;

    for (auto n:nums1)
        n1.insert(n);
    
    for (auto n:nums2)
        n2.insert(n);

    for (auto n:n1){
        if (n2.find(n)!=n2.end())
            _intersection.push_back(n);
    }

    return _intersection;
    // todo: optimize with n+m
}

int main()
{
    vector<int> nums1 = {1, 2, 2, 1};
    vector<int> nums2 = {2, 2};
    intersection(nums1,nums2);
    return 0;
}